File: Data Structures & Algorithms/maximum-product-subarray/submission-4.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res = INT_MIN;

        int currMax = 1;
        int currMin = 1;
        for (int i = 0; i < nums.size(); ++i) {
            int nextMax = max(currMax * nums[i], currMin * nums[i]);
            int nextMin = min(currMax * nums[i], currMin * nums[i]);

            currMax = max(nextMax, nums[i]);
            currMin = min(nextMin, nums[i]);

            res = max(currMax, res);
        }

        return res;
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-4.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        res = [1 for _ in range(len(nums))]
        maxRes = 1

        for i in range(len(nums) - 1, -1, -1):
            for j in range(i + 1, len(nums)):
                if nums[j] > nums[i]:
                    res[i] = max(res[i], 1 + res[j])
            
            maxRes = max(res[i], maxRes)

        return maxRes

